### PR TITLE
Implementation of project board automations

### LINKED ISSUE #3374: Implementation of the project automation changes
The implementation plan for the project board automations was merged in #3042.

The goal of the issue is to implement the changes outlined in the [plan](https://docs.openverse.org/projects/proposals/project_improvement/20230914-implementation_plan_project_automations.html).

### PR DESCRIPTION
**Pull request by dhruvkb**
## Fixes

Fixes #3374 by @dhruvkb 

## Description

This PR implements the changes and additions to the project automations as per the plan outlined in #3042.

Given how GitHub Actions development goes, I'm anticipating quite a few revisions on this work once the actions start running and failing and exposing edge cases so if you could review it a bit deeply, we might catch and patch those cases faster.

## Testing Instructions

The changes are so tightly coupled with GitHub that it's very hard to test them at all. Also they need an access token making it not possible to test them in CI. However here are some bits that you can test locally.

### `Project`

1. Create and export your personal access token `export ACCESS_TOKEN=<your token>`.
2. Open `node` REPL from the `automations/js` directory.
3. Try this:
   ```node
   > const octokit = require('@octokit/rest')
   > const projects = await import('./src/utils/projects.mjs')
   >
   > const { ACCESS_TOKEN } = process.env
   > const oct = new octokit.Octokit({ auth: ACCESS_TOKEN })
   > const backlogBoard = await projects.getBoard(oct, 'Backlog')
   > backlogBoard.projectId
     'PVT_kwDOAAQ2Js4AMZdL'
   > backlogBoard.columns
     {
       Blocked: '⛔ Blocked',
       Backlog: '📋 Backlog',
       ToDo: '📅 To Do',
       InProgress: '🏗 In Progress',
       Done: '✅ Done',
       Discarded: '🗑 Discarded'
     }
   ```

### `PullRequest`

1. Create and export your personal access token `export ACCESS_TOKEN=<your token>`.
2. Open `node` REPL from the `automations/js` directory.
3. Try this:
   ```node
   > const octokit = require('@octokit/rest')
   > const pr = await import('./src/utils/pr.mjs')
   >
   > const { ACCESS_TOKEN } = process.env
   > const oct = new octokit.Octokit({ auth: ACCESS_TOKEN })
   > let pull = new pr.PullRequest(oct, "PR_kwDOFU_ugM5f4yt7")
   > await pull.init()
   > pull.reviewDecision
     'CHANGES_REQUESTED'
   > pull.linkedIssues
     [ 'I_kwDOFU_ugM5ssiSa' ]
   > pull.reviewCounts
	 {
	   APPROVED: 0,
	   COMMENTED: 1,
	   CHANGES_REQUESTED: 1,
	   DISMISSED: 0,
	   PENDING: 0
	 }
   ```

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: automations/js/src/utils/projects.mjs [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+      break
+    case 'Discussions':
+      project = new Project(octokit, 'WordPress', 79)
+      break
+    case 'Project Tracker':
+      project = new Project(octokit, 'WordPress', 70)
+      break
+    default:
+      throw new Error(`Unknown project board "${name}".`)
+  }
```
Comment (sarayourfriend) [inline automations/js/src/utils/projects.mjs:226]: nit: I think this can be significantly simplified and made easier to update by using an object to configure the mapping of project name to id:

```suggestion
  const projects = {
    Backlog: 75,
    PRs: 98,
    Todos: 59,
    Discussions: 79,
    "Project Tracker": 70,
  }
  
  if (!projects[name]) {
    throw new Error(`Unknown project board "${name}".`)
  }
  
  const project = new Project(octokit, 'WordPress', projects[name])
```

This makes adding/removing projects a bit easier, just one line to add instead of 3, and makes the relationship of each case statement clearer (it's really just 1-to-1 project name to project id).

File: automations/js/src/project_automation/prs.mjs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      break
+
+    case 'ready_for_review':
+      await syncReviews()
+      break
+
+    case 'closed':
+      if (!eventPayload.pull_request.merged)
+        await syncIssues(backlogColumns.Backlog)
+      break
```
Comment (sarayourfriend) [inline automations/js/src/project_automation/prs.mjs:85]: nit (and should be an eslint rule, really) but can you add braces around the cases? For longer cases it makes it easier to read and harder to accidentally forget a break/return or lose track of fall-through. Should all be eslint rules, though.

File: automations/js/src/project_automation/prs.mjs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  const reviewDecision = pr.reviewDecision
+  const reviewCounts = pr.reviewCounts
+
+  if (reviewDecision === 'APPROVED')
+    await prBoard.moveCard(card.id, columns.Approved)
+  else if (reviewDecision === 'CHANGES_REQUESTED')
+    await prBoard.moveCard(card.id, columns.ChangesRequested)
+  else if (reviewCounts.APPROVED === 1)
+    await prBoard.moveCard(card.id, columns.Needs1Review)
+  else await prBoard.moveCard(card.id, columns.Needs2Reviews)
```
Comment (sarayourfriend) [inline automations/js/src/project_automation/prs.mjs:45]: Same nit here, really, just to add braces. One-line ifs seem cool but the second there's a need to add another statement to any condition, it's more work and easy to mess up by forgetting to add the braces and causing weird, hard to find bugs.

File: .github/workflows/pr_automations.yml [hunk 4]
```diff
@@ -0,0 +1,78 @@
+# Runs all automations related to PR events.
+#
+# See `issue_automations_init.yml` for the corresponding implementation for
+# issues.
```
Comment (sarayourfriend) [inline .github/workflows/pr_automations.yml:4]: Does this file exist? If not, we can remove this comment

```suggestion
```

Oh actually, seeing that the same comment is in the other pr automations workflow file. Is this meant to be this instead?

```suggestion
#
# See `issue_automations.yml` for the corresponding implementation for issues.
```

File: .github/workflows/pr_automations_init.yml [hunk 5]
```diff
@@ -0,0 +1,54 @@
+# Initialises all automations related to PR events.
+#
+# See `issue_automations_init.yml` for the corresponding implementation for
+# issues.
```
Comment (sarayourfriend) [inline .github/workflows/pr_automations_init.yml:4]: If the file doesn't exist, we can remove this comment:

```suggestion
```

**Review by sarayourfriend — APPROVED** [review]
LGTM! Left some nit-picky comments but nothing blocking. I didn't carefully review all the implementation details, they all look good. The flows make sense to me, at least. To make testing easier (for example, if we need a fast-follow to fix some bugs) we could add `workflow_dispatch` to the new workflows with an input for passing a custom event configuration. In any future PRs, maintainers can then dispatch the workflow on the branch to test changes. Not sure totally how that would work with the workflow_run event though. Just a thought :grin:

File: automations/js/src/utils/projects.mjs (hunk 1 above)
Comment (dhruvkb) [reply automations/js/src/utils/projects.mjs]: This is really great, and should've been the first solution I thought of. I don't even remember why I used `switch`-`case` in the first place 🤦. Fixed in aaf807f.

File: automations/js/src/project_automation/prs.mjs (hunk 2 above)
Comment (dhruvkb) [reply automations/js/src/project_automation/prs.mjs]: Fixed in f4ba8ef.

File: .github/workflows/pr_automations_init.yml (hunk 5 above)
Comment (dhruvkb) [reply .github/workflows/pr_automations_init.yml]: Fixed in 53142df. The file exists, the `_init.yml` suffix is incorrect.

File: .github/workflows/pr_automations.yml (hunk 4 above)
Comment (dhruvkb) [reply .github/workflows/pr_automations.yml]: Fixed in 53142df. The file exists, the `_init.yml` suffix is incorrect.

File: automations/js/src/project_automation/prs.mjs (hunk 3 above)
Comment (dhruvkb) [reply automations/js/src/project_automation/prs.mjs]: Added braces in 53142df.

File: automations/js/src/utils/octokit.mjs [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+}
+
+/**
+ * Get an authenticated instance of `Octokit` to make GitHub API calls.
+ *
+ * @returns {Octokit} an authenticated instance of `Octokit`
+ */
+export function getOctokit() {
+  return new Octokit({ auth: pat })
+}
```
[the remark this answers is not in the document]
Comment (dhruvkb) [reply automations/js/src/utils/octokit.mjs:17]: I'll see if using `actions/github-script` can help here. I'm not too sure (especially for the PRs side which runs using `workflow-run` and thus needs context to be provided to it via artifacts) but it will be interesting to explore that.

[the remark this answers is not in the document]
Comment (dhruvkb) [reply automations/js/src/utils/octokit.mjs]: I converted the script to `actions/github-script` to explore and it has had the following impact.

- Changes to surrounding code were reduced, meaning 8 fewer files to review.
- The new files themselves became a bit more complex to be able to be referenced from the action.

Let me know if you think this is a net improvement or if we should go back to aaf807f.

[the remark this answers is not in the document]
Comment (sarayourfriend) [reply automations/js/src/utils/octokit.mjs]: To me, it does read as an improvement, and I don't find it harder to reference from the action, but do acknowledge it makes it much harder (maybe just impossible) to run them locally. https://github.com/nektos/act could be a workaround for that, potentially even a better approach because it doesn't rely on manually configuring event payloads, which would be easy to get incorrect. We don't have good ways to test any other part of our CI workflows, so these ones wouldn't be unique.

I think it basically has a net-zero effect and either approach is fine, so totally your preference. My approval of the original one was genuine :slightly_smiling_face:

[the remark this answers is not in the document]
Comment (dhruvkb) [reply automations/js/src/utils/octokit.mjs]: In that case, let's stick with the changes. Considerable brainpower and time had to be invested (as must be evident from the commits) to get to where it is. Thanks for the pointer, it reduced the changed files by a third!

File: .github/workflows/pr_automations.yml [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+               artifact_id: matchArtifact.id,
+               archive_format: 'zip',
+            });
+            let fs = require('fs');
+            fs.writeFileSync(`${process.env.GITHUB_WORKSPACE}/event_info.zip`, Buffer.from(download.data));
+
+      - name: Unzip artifact
+        run: |
+          unzip event_info.zip
+          mv event.json /tmp/event.json
```
Comment (dhruvkb) [inline .github/workflows/pr_automations.yml:72]: This is one thing that's bothering me a lot, it's two lines of Bash breaking up two `action/github-script`s. Surely it could be done inside JS? Or the previous step in Bash?

File: automations/js/src/utils/pr.mjs [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  /**
+   * Get additional information about the PR such as the linked issues and the
+   * review decision, as well the states of all submitted reviews.
+   *
+   * @returns {Promise<PrDetails>}
+   */
+  async getPrDetails() {
+    const res = await this.octokit.graphql(
+      `query getPrDetails($id: ID!) {
+        node(id: $id) {
```
Comment (krysal) [inline automations/js/src/utils/pr.mjs:44]: Where does the node ID come from here❓

**Review by krysal — APPROVED** [review]
This looks excellent, @dhruvkb! Nice finding a workaround for PRs. Given the complexity, it looks like that could have been a separate PR, but now it's done 👍

File: automations/js/src/utils/pr.mjs (hunk 8 above)
Comment (dhruvkb) [reply automations/js/src/utils/pr.mjs:44]: It is passed as a parameter on L63.

Comment (krysal) [reply automations/js/src/utils/pr.mjs:44]: Thanks for the explanation! I somehow skipped the reading of the file so it's much more clear now 😄
